### PR TITLE
cmakeBuild: add support for targets with dependencies

### PR DESCRIPTION
**Pull request by alexandrumc**
Depends on https://github.com/atilaneves/reggae/pull/251

### CODE REVIEW

File: payload/reggae/build.d [hunk 1]
```diff
@@ -598,9 +598,9 @@ struct Command {
 
             case link: {
                 version(Windows)
-                    auto cArgs = ["/nologo", "/Fe$out", "$flags", "$in"];
+                    auto cArgs = ["/nologo", "/Fe$out", "$flags", "$link_libraries", "$in"];
                 else
-                    auto cArgs = ["-o", "$out", "$flags", "$in"];
+                    auto cArgs = ["$flags", "$in", "-o", "$out", "$link_libraries"];
```
Comment (atilaneves) [inline payload/reggae/build.d:603]: Why the change in order?

File: payload/reggae/rules/cmake.d [hunk 2]
```diff
@@ -26,24 +27,36 @@ imported!"reggae.build".Target[] cmakeBuild(imported!"reggae.types".ProjectPath
     const configuration = configurations.front;
 
     const cmakeTargetEntries = configuration["targets"].array;
+    foreach (targetEntry; cmakeTargetEntries) {
```
Comment (atilaneves) [inline payload/reggae/rules/cmake.d:30]: Could you explain please what's going on here exactly?

File: payload/reggae/rules/cmake.d [hunk 3]
```diff
@@ -161,14 +174,26 @@ private bool isHeaderFile(in string file) {
...
 
-private string getLinkerFlags(in imported!"std.json".JSONValue target) {
+private struct CMakeLinkerFlags {
+    import reggae.types : Flags, LibraryFlags;
+    Flags flags;
+    LibraryFlags libraryFlags;
+    Flags libraryPathFlags;
+
+    Flags getRegularFlags() const {
+        return Flags(flags.value.dup ~ libraryPathFlags.value.dup);
```
Comment (atilaneves) [inline payload/reggae/rules/cmake.d:184]: Do they need to be duplicated?

File: payload/reggae/rules/cmake.d [hunk 4]
```diff
@@ -161,14 +174,26 @@ private bool isHeaderFile(in string file) {
...
+private CMakeLinkerFlags getLinkerFlags(in imported!"std.json".JSONValue target, ref CMakeInfo cmakeInfo) {
+    import reggae.types : Flags, LibraryFlags;
     import std.array : join;
     import std.algorithm : each;
 
     string flags, libraries, libraryPath;
 
     if ("link" !in target) {
-        return null;
+        return CMakeLinkerFlags.init;
```
Comment (atilaneves) [inline payload/reggae/rules/cmake.d:196]: `CmakeLinkerFlags()`?

File: payload/reggae/rules/cmake.d [hunk 5]
```diff
@@ -184,7 +209,12 @@ private string getLinkerFlags(in imported!"std.json".JSONValue target) {
                 flags ~= cf["fragment"].str ~ " ";
                 break;
             case "libraries":
-                libraries ~= cf["fragment"].str ~ " ";
+                const libName = cf["fragment"].str;
+                string libPath = libName;
+                if (auto p = libName in cmakeInfo.cmakeToReggaeArtifactPath) {
+                    libPath = *p;
```
Comment (atilaneves) [inline payload/reggae/rules/cmake.d:215]:
```d
const libPath = () {
    if (auto p = libName in cmakeInfo.cmakeToReggaeArtifactPath) 
        return *p;
    return cf["fragment"].str;
}();
```

**Review by atilaneves — COMMENTED** [review]
This seems like a lot for one commit; I'm having difficulties understanding why the various changes are necessary given the PR title.

File: payload/reggae/rules/cmake.d (hunk 3 above)
Comment (alexandrumc) [reply payload/reggae/rules/cmake.d]: I was trying to avoid the 

> Error: `inout` constructor `reggae.types.Flags.this` creates const object, not mutable

Deleted this method and moved the work to the caller function.

File: payload/reggae/build.d (hunk 1 above)
Comment (alexandrumc) [reply payload/reggae/build.d:603]: Basically, the answer is "that's how Ninja does it". 

With the _old_ order, the problem that I encountered had to do with the fact that the static library was _not_ the last thing in the command line, e.g. `gcc -o binary flag1 flag2 libStatic.a main.o`. When this happened, it did not link correctly and complained that the symbols referenced by `main.o` couldn't be found. Note that on my machine (I'm using Ubuntu 20) the test was passing and there were no problems. However, on these upstream Linux runners (that use Ubuntu 22) the test was failing, so I changed the order and it worked.

As I said, that's also how Ninja does it. And, in general, what I've seen is that Ninja makes a lot of distinctions when it comes to `flags`, which reggae doesn't when it generates the `build.ninja`  as it just dumps everything under the same `flags` tag and is not interested in the order of the flags at all. On the other hand, Ninja makes all these distinctions between the flags essentially to pass everything in the right order.

However, for the moment, I think what we have suits our needs.

**Review by atilaneves — APPROVED** [review]
